datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -285,57 +287,57 @@ impl LogicalPlan {
     /// Calls `f` on all expressions (non-recursively) in the current
     /// logical plan node. This does not include expressions in any
     /// children.
-    pub fn inspect_expressions<F, E>(self: &LogicalPlan, mut f: F) -> Result<(), E>
-    where
-        F: FnMut(&Expr) -> Result<(), E>,
-    {
+    pub fn apply_expressions<F: FnMut(&Expr) -> Result<TreeNodeRecursion>>(
```

Comment:
This is an api change as it renames a public function. I couldn't find a way to get the error type to line up, so I think it is unavoidable.

I'll mark this PR as API change

Reply:
Actually, I found another pattern here: https://github.com/apache/arrow-datafusion/blob/d201ec742b08fa2a0549fc27f0b39622d67391ab/datafusion/expr/src/utils.rs#L684

Maybe we can pull this part (add `apply_expressions` into a separate PR)? I think it would be valuable in its own -- for example it could be used to void the `expressions()` copy here:

https://github.com/apache/arrow-datafusion/blob/27bdf3c8c976874aae2417cb2aa435448832ecc0/datafusion/optimizer/src/analyzer/subquery.rs#L236-L241

It could also make this PR smaller

Reply:
Yes, we can do that. Let me file a new PR with this change tomorrow and then look into the infinite recursion issue of this PR (https://github.com/peter-toth/arrow-datafusion/pull/3). Until then I'm marking this PR as draft.

Reply:
I've opened https://github.com/apache/arrow-datafusion/pull/9984 as I think we don't need that `inspect_expr_pre()`, this PR together with https://github.com/apache/arrow-datafusion/pull/9984 will remove all usecases of it.

Reply:
As for https://github.com/apache/arrow-datafusion/blob/27bdf3c8c976874aae2417cb2aa435448832ecc0/datafusion/optimizer/src/analyzer/subquery.rs#L236-L241 I plan to check all `.expressions()` calls and see if it can be transformed to the better `.apply_expressions()` once this PR is merged.

Reply:
Sounds good

Reply:
> I've opened #9984 as I think we don't need that `inspect_expr_pre()`. This PR together with #9984 will remove all usecases of it.

I merged #9984

In order to make this PR easier to merge (so it isn't an API change) I 
1. pushed 7c90e70dc to restore `inspect_expressions` (by calling `apply_expressions`) and deprecated it. 
2. Removed the API change

Reply:
> I plan to check all .expressions() calls and see if it can be transformed to the better .apply_expressions() once this PR is merged.

This would be good -- I think there are some more (though https://github.com/apache/arrow-datafusion/pull/9974 did get merged)